### PR TITLE
Build + Imports + Source refactors

### PR DESCRIPTION
**Pull request by lasote**
#1823 (#1818 and #1819 issues too)

Also including  #1860 #1865 (closed, because have been refactored together)

Question, I think we should remove the `--cwd` from conan install in favor of a `--build_folder`. Because of two reasons: consistency with all the new commands and the concern of what's exactly the meaning of `cwd` and how affects the rest of the paths specified in the command line.

### CODE REVIEW

File: conans/client/conan_api.py [hunk 1]
```diff
@@ -544,24 +539,67 @@ def package(self, reference="", package_id=None, build_folder=None, source_folde
             self._manager.local_package(package_folder, recipe_folder, build_folder, source_folder)
 
     @api_method
-    def source(self, reference, force=False, cwd=None):
-        cwd = prepare_cwd(cwd)
-        current_path, reference = _get_reference(reference, cwd)
-        self._manager.source(current_path, reference, force)
+    def source(self, path, source_folder=None, build_folder=None):
+        path = self._abs_relative_to(path, os.getcwd())
+        source_folder = self._abs_relative_to(source_folder, os.getcwd()) or os.getcwd()
```
Comment (memsharded) [inline conans/client/conan_api.py]: A bit ugly so much ``os.getcwd()``. Maybe add an argument to the ``abs_relative_to``, that is defaulted to None, and could be passed the cwd:

```python
cwd = os.getcwd()
path = self._abs_relative_to(path, cwd)
source_folder = self._abs_relative_to(source_folder, cwd, default_dir=cwd)
```

File: conans/client/conan_api.py [hunk 2]
```diff
@@ -544,24 +539,67 @@ def package(self, reference="", package_id=None, build_folder=None, source_folde
             self._manager.local_package(package_folder, recipe_folder, build_folder, source_folder)
 
     @api_method
-    def source(self, reference, force=False, cwd=None):
-        cwd = prepare_cwd(cwd)
-        current_path, reference = _get_reference(reference, cwd)
-        self._manager.source(current_path, reference, force)
+    def source(self, path, source_folder=None, build_folder=None):
```
Comment (memsharded) [inline conans/client/conan_api.py]: Do we want to directly break? Not showing a warning for the ``conan source <pkg_ref>`` command?

File: conans/test/command/source_test.py [hunk 3]
```diff
@@ -1,44 +1,37 @@
...
-        self.assertNotIn("Hello/0.1@lasote/stable: Running source!", client.user_io.out)
-
-        # Forced should have effect
-        client.run("source Hello/0.1@lasote/stable --force")
-        self.assertIn("WARN: Forced removal of source folder", client.user_io.out)
-        self.assertIn("Hello/0.1@lasote/stable: Configuring sources", client.user_io.out)
-        self.assertIn("Hello/0.1@lasote/stable: Running source!", client.user_io.out)
+        subdir = os.path.join(client.current_folder, "subdir")
+        os.mkdir(subdir)
+        client.run("install .. --cwd subdir")  # IMPORTANT! SHOULD WE CHANGE THE CWD FROM INSTALL TO BUILD_FOLDER?
```
Comment (memsharded) [inline conans/test/command/source_test.py]: Yes, sounds we should

File: conans/client/conan_api.py (hunk 2 above)
Comment (lasote) [reply conans/client/conan_api.py]: Look at command to see the check

**Pull request merged**